File: 2025-04-07/forkjoin/src/test/java/SumForkJoinTest.java
import org.junit.Test;

import java.util.Random;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

public class SumForkJoinTest {

    private double sequentialTotal(int[] arr) {
        double sum = 0;
        for (int i=0; i < arr.length; i++) {
            sum += Math.sin((arr[i]));
        }
        return sum;
    }

    @Test
    public void testSum() throws Exception {
        Random r = new Random(90125);
        int[] arr = new int[10_000_000];
        for (int i = 0; i < arr.length; i++) {
            arr[i] = r.nextInt();
        }
        // Do the sequential test a number of times to warm it up, measure on
        // last
        int numWarmups = 3;
        long timeBefore = 0;
        long seqTime = 0;
        double seqSum = 0;
        for (int i=0; i < numWarmups; i++) {
            timeBefore = System.currentTimeMillis();
            seqSum = sequentialTotal(arr);
            seqTime = System.currentTimeMillis() - timeBefore;
        }

        double parallelSum = 0;
        long parallelTime = 0;

        // Do the parallel test a number of times to warm it up, measure on last
        // int[] seqCutoffs = {1, 2, 10, 100, 1000, 10000};
        int[] seqCutoffs = {
            1, 2, 3, 4, 10, 100, 1000, 10000, 100_000, 1_000_000, 5_000_000
        };
        System.out.println("cutoff / seq time / parallel time / speedup");
        for (int seqCutoff : seqCutoffs) {
            // Do the parallel test a number of times to warm it up, measure on
            // last
            for (int i = 0; i < numWarmups; i++) {
                timeBefore = System.currentTimeMillis();
                parallelSum = SumForkJoin.sum(arr, seqCutoff);
                parallelTime = System.currentTimeMillis() - timeBefore;
            }
            assertEquals(seqSum, parallelSum, 1e-9);
            double speedup = ((double) seqTime) / parallelTime;
            System.out.printf("%10d %5d %5d %7.3f\n", seqCutoff, seqTime,
                              parallelTime, speedup);
        }
    }
}

File: 2025-04-07/forkjoin/src/test/kotlin/SumForkJoinTest.kt
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue
import kotlin.random.Random
import kotlin.time.measureTimedValue


class SumForkJoinTest {

    fun sequentialTotal(arr: List<Int>): Int {
        var sum = 0
        for (i in 0..<arr.count()) {
            sum += if (isPrime(arr[i])) 1 else 0
        }
        return sum
    }

    @Test
    fun testSum() {
        var r = Random(90125)
        val arr = mutableListOf<Int>()
        for (i in 0..<100_000) {
            arr.add(r.nextInt(10000))
        }

        // Do the sequential test a number of times to warm it up, then do it
        // one more time and measure on last
        val numWarmups = 3
        // total each time; display it to make sure it doesn't get
        // optimized out
        var total = 0.0
        for (i in 0..<numWarmups) {
            total += sequentialTotal(arr)
        }
        val (seqSum, seqTime) = measureTimedValue {
            sequentialTotal(arr)
        }
        // Print the warmup total just to make sure that the code
        // doesn't get optimized out
        println("Warmup sequential total = ${total}")
        println("Sum of list = ${seqSum}")
        println("Measured sequential time = ${seqTime}")

        val numProcessors = Runtime.getRuntime().availableProcessors()
        println("Number of processors = ${numProcessors}")
        println("-----------------------------")
        val seqCutoffs = listOf(
            1, 2, 3, 4, 10, 100, 1000, 10_000, 50_000, 100_000
        )
        println("cutoff / seq time / parallel time / speedup")

        for (seqCutoff in seqCutoffs) {
            // Do the parallel test a number of times to warm it up, measure on
            // last
            total = 0.0
            for (i in 0..<numWarmups) {
                total += sumMultiThreaded(arr, seqCutoff)
            }
            val (parallelSum, parallelTime) = measureTimedValue {
                sumMultiThreaded(arr, seqCutoff)
            }

            val speedup = seqTime/parallelTime
            println(String.format("%10d %s %s %7.3f", seqCutoff, seqTime,
                                  parallelTime, speedup))
            assertEquals(seqSum, parallelSum)
        }
    }
}
